### ISSUE TITLE
Problem: Archivematica's mechanism to import custom structmaps can only describe flat structures, e.g. no nested-dirs, and only for files with limited character-sets, e.g. not even spaces are allowed

### ISSUE THREAD

**Issue by ross-spencer** [archivematica/Issues#283]
**Expected behaviour**

The ability to supply a custom `structmap` that describes logical structure of arbitrary complexity.

**Current behaviour**

The current behavior is that the template for providing a custom `structmap` uses a `FILEID` field in an `fptr` to describe unique filenames likely to appear in a transfer. These values are replaced with `filesec` UUIDs through a lookup when the AIP METS is written. 

Ref: https://github.com/artefactual/archivematica-sampledata/blob/f35f041af690774653b63fcd94e6649e719fed66/SampleTransfers/structMapSample/metadata/mets_structmap.xml

Unfortunately, the `FILEID` field is restricted to a specific character set which means that the slashes normally used to describe nested directories are considered to be illegal: https://stackoverflow.com/questions/1631396/what-is-an-xsncname-type-and-when-should-it-be-used/6159003#6159003 

When a user attempts to supply a custom structmap during transfer. and if that structmap attempts to define some structure, e.g. `some/path/to/file.txt` then the validation of that structmap will fail and the transfer cannot continue.

**Proposed solution**

A solution has been proposed to use the `CONTENTIDS` field of an `fptr` as this can encode a URI, for example, a `FILE:///` uri. Though we elected to use a far simpler path-like string here:

```
<mets:mets xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:mets="http://www.loc.gov/METS/" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.loc.gov/METS/ http://www.loc.gov/standards/mets/mets.xsd" OBJID="ARCH00152.355">
  <mets:structMap TYPE="logical" LABEL="IISH Sequence">
    <mets:div TYPE="brochure">
      <mets:div TYPE="page" ORDER="1" LABEL="Page 1">
        <mets:fptr FILEID="one_file_two.txt" CONTENTIDS="sample_dir/one_file_two.txt" />
      </mets:div>
    </mets:div>
  </mets:structMap>
</mets:mets>
```

The code-change for this is fairly minimal, **nb** this is just a work in progress commit: https://github.com/IISH/archivematica/commit/c7cc6cf41a03755972169729f2db9efb85161fb4

The change reads the `CONTENTIDS` field to then does the lookup. The `FILEID` field continues to be replaced with the file UUIDs found in the SIP. 

**Other considerations**

Given this is a URI field, we might find that it won't validate given certain combinations of characters. Therefore we may consider not keeping the `CONTENTIDS` field in the XML once  it is written (where the current implementation doesn't have these at all). Or we need to consider the impact of URI encoding on making this as easy as possible for users. 

**Your environment (version of Archivematica, OS version, etc)**

`qa/1.x`.

**Additional context**

Duplicate of: https://github.com/IISH/archivematica/issues/4 
Related to https://github.com/IISH/archivematica/issues/5 
Related to https://github.com/archivematica/Issues/issues/245 

---
**For Artefactual use:**
Please make sure these steps are taken before moving this issue from Review to Verified in Waffle:

- All PRs related to this issue are properly linked 👍
- All PRs related to this issue have been merged 👍
- Test plan for this issue has been implemented and passed 👍
- Documentation regarding this issue has been written and it has been added to the release notes, if needed 👍

**Comment by ross-spencer** [comment]
TODO:

* [ ] Lookup expression to understand why `<!-- File: 2017-01-27-with-good-reason-ferdinand-short.mp3 -->` fails.

**Comment by ross-spencer** [comment]
**Analysis**

Archivematica uses [this script](https://github.com/artefactual/archivematica/blob/stable/1.9.x/src/MCPClient/lib/clientScripts/archivematicaVerifyMets.sh) to validate a structmap supplied with a transfer.

It calls the following:
```bash
if [ -f "${metsFile}" ]; then
    xmllint --noout --schema "${schemaFile}" "${metsFile}"
else
    echo "No metadata/mets_structmap.xml file to verify."
fi
```

Using this we can rig up a test file with test patterns in it, designed to match a XSD which will do something similar to a mets schema. 

For the two sample files described below we can generate the following errors: 
```bash
$ xmllint --schema structmap-simulation.xsd field-examples.xml --noout
field-examples.xml:3: element FILEID: Schemas validity error : Element 'FILEID': '2019-04-02-transfer.file' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:4: element FILEID: Schemas validity error : Element 'FILEID': '-transfer file.zip' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:5: element FILEID: Schemas validity error : Element 'FILEID': '.transfer-file.zip' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:6: element FILEID: Schemas validity error : Element 'FILEID': '1-transfer' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:9: element FILEID: Schemas validity error : Element 'FILEID': '1transfer' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:10: element FILEID: Schemas validity error : Element 'FILEID': '2600.zip' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:12: element FILEID: Schemas validity error : Element 'FILEID': 'a@' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:13: element FILEID: Schemas validity error : Element 'FILEID': 'has#' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:14: element FILEID: Schemas validity error : Element 'FILEID': 'handle/1234' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:15: element FILEID: Schemas validity error : Element 'FILEID': '001-digitised.image.tif' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml:16: element FILEID: Schemas validity error : Element 'FILEID': 'this/is/a/nested/dir/file.txt' is not a valid value of the atomic type 'xs:IDREF'.
field-examples.xml fails to validate
```

Note how they are all for the `FILEID` and not for the `CONTENTIDS`. 

*Sample files:*

`mets_structmap.xml`
```xml
<?xml version="1.0" encoding="utf-8"?>
<mets:mets xmlns:mets="http://www.loc.gov/METS/">
  <mets:structMap TYPE="logical">
    <mets:div LABEL="Short Listen: Ferdinand, the Misunderstood Bull by With Good Reason" TYPE="documentary">
      <mets:div LABEL="Introduction" ORDER="1">
        <!-- File: 2017-01-27-with-good-reason-ferdinand-short.mp3 -->
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:00" END="00:00:17" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="The story of Ferdinand and its context in history" ORDER="2">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:18" END="00:01:13" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Sharon McQueen - Socio Cultural Historian of Children's Literature" ORDER="3">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:01:14" END="00:01:33" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Isolation and Pacifism" ORDER="4">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:01:33" END="00:02:09" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Written on a rainy afternoon by Munro Leaf" ORDER="5">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:02:10" END="00:02:52" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Individualism" ORDER="6">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:02:53" END="00:03:33" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Banned in Germany, Spain, and attempted in the USA" ORDER="7">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:03:34" END="00:04:05" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Ferdinand's fame, even as he's seen as a 'coward'" ORDER="8">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:04:06" END="00:04:35" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Outro" ORDER="9">
        <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:00" END="00:05:06" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <!-- documentary -->
    </mets:div>
  </mets:structMap>
</mets:mets>
```

`structmap-simulation.xsd`
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
	<xs:element name="structmapfields">
		<xs:complexType>
			<xs:sequence>
				<xs:element name="FILEID" type="xs:IDREF" minOccurs ="0" maxOccurs="unbounded" />
				<xs:element name="CONTENTIDS" type="URIs" minOccurs ="0" maxOccurs="unbounded" />
			</xs:sequence>
		</xs:complexType>
	</xs:element>

	<!-- from mets.xsd -->
	<xs:simpleType name="URIs">
	    <xs:list itemType="xs:anyURI"/>
	</xs:simpleType>

</xs:schema>


<!-- from mets.xsd:

<xsd:attribute name="FILEID" type="xsd:IDREF" use="optional">
	<xsd:annotation>
		<xsd:documentation xml:lang="en">FILEID (IDREF/O): An optional attribute that provides the XML ID identifying the &lt;file&gt; element that links to and/or contains the digital content represented by the &lt;fptr&gt;. A &lt;fptr&gt; element should only have a FILEID attribute value if it does not have a child &lt;area&gt;, &lt;par&gt; or &lt;seq&gt; element. If it has a child element, then the responsibility for pointing to the relevant content falls to this child element or its descendants.
		</xsd:documentation>
	</xsd:annotation>
</xsd:attribute>
<xsd:attribute name="CONTENTIDS" type="URIs" use="optional">
    <xsd:annotation>
    	<xsd:documentation xml:lang="en">CONTENTIDS (URI/O): Content IDs for the content represented by the &lt;fptr&gt; (equivalent to DIDL DII or Digital Item Identifier, a unique external ID).
        </xsd:documentation>
    </xsd:annotation>
    </xsd:attribute>

-->
```

**Comment by ross-spencer** [comment]
Additionally, using the example described in the [METS documentation](https://www.loc.gov/standards/mets/METSOverview.v2.html#structmap), we cannot populate all the FILEIDs as anticipated:
```xml
<mets:structMap TYPE="logical" ID="structMap_2">
    <mets:div TYPE="track" LABEL="Complete documentary">
      <!-- File: ferdinand_short_2017_01_27.mp3

          NB. The filename is designed to fit the original design of the
          custom structMap functionality and so it is kept as simple as
          possible and will validate as a xs:IDREF'.
        -->
      <mets:div LABEL="Introduction" ORDER="1">
        <mets:fptr FILEID="file-78a8587a-1788-4a9f-91d9-6cf3e442ced6">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:00" END="00:00:17" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
      <mets:div LABEL="Outro" ORDER="2">
        <mets:fptr FILEID="file-78a8587a-1788-4a9f-91d9-6cf3e442ced6">
          <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:18" END="00:01:13" BETYPE="TIME"/>
        </mets:fptr>
      </mets:div>
    </mets:div>
  </mets:structMap>
</mets:mets>
```

**Comment by evelynPM** [comment]
I ingested a transfer with the sample "Ferdinand" structMap supplied above. The resulting AIP METS file fails validation with the error "Cvc-id.1: There Is No ID/IDREF Binding For IDREF 'ferdinand_short_2017_01_27.mp3'., Line '298', Column '13'." Presumably that's because IDREF isn't referencing a corresponding entry in the fileSec.

**Comment by ross-spencer** [comment]
Hi there @evelynPM yeah, I think that's the only reason. 

I created some sample data [here](https://github.com/artefactual/archivematica-sampledata/tree/master/SampleTransfers/StructMapTransferSamples) the Ferdinand example is under [SingleObjectExample](https://github.com/artefactual/archivematica-sampledata/tree/master/SampleTransfers/StructMapTransferSamples/SingleObjectExample) if you'd like to review these. A cursory review on my machine shows those samples should pass validation okay.

**Comment by ross-spencer** [comment]
> I ingested a transfer with the sample "Ferdinand" structMap supplied above. The resulting AIP METS file fails validation with the error "Cvc-id.1: There Is No ID/IDREF Binding For IDREF 'ferdinand_short_2017_01_27.mp3'., Line '298', Column '13'." Presumably that's because IDREF isn't referencing a corresponding entry in the fileSec.

Ah, I see what you mean. I think we can make this more robust if that's happening. I'll see if I can recreate and see what's needed to do this...

**Comment by ross-spencer** [comment]
@evelynPM I've submitted a draft PR [here](https://github.com/artefactual/archivematica/pull/1455)

We can output something like this in the microservice job:
```
Skipping creation of normative structmap
Deleted empty directory /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/metadataReminder/122-c90f1b96-f6ce-4805-9f8c-41cd5f744b06/logs/transfers/122-5920296a-6e93-421b-9b76-46f451afb475/logs/fileMeta
Deleted empty directory /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/metadataReminder/122-c90f1b96-f6ce-4805-9f8c-41cd5f744b06/logs/fileMeta
Deleted empty directory /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/metadataReminder/122-c90f1b96-f6ce-4805-9f8c-41cd5f744b06/thumbnails
--->   No CONTENTIDS found in custom structmap. Structmap will not be included in AIP METS   <--- structmap warning
DmdSecs: 1
AmdSecs: 5
TechMDs: 5
RightsMDs: 0
DigiprovMDs: 35
```
Would that be okay? And my preference would be to not let the process fail at this point, but we haven't many ways to raise that as a warning. Do you have an opinion?

The reason the structmap example above doesn't work is that we changed:
```xml
     <mets:fptr FILEID="ferdinand_short_2017_01_27.mp3">
       <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:00" END="00:00:17" BETYPE="TIME"/>
     </mets:fptr>
```
To require:
```xml
     <mets:fptr FILEID="some_value" CONTENTID="ferdinand_short_2017_01_27.mp3">
       <mets:area FILEID="ferdinand_short_2017_01_27.mp3" BEGIN="00:00:00" END="00:00:17" BETYPE="TIME"/>
     </mets:fptr>
```
Which you'll see in the sample data linked to. It looks like we missed this test case in between the original and new versions of this capability.

**Comment by evelynPM** [comment]
I think using CONTENTIDS on user-supplied structMaps if the content is not in a flat file structure is ok. If the content is in a flat structure, the user should not be required to include CONTENTIDS in the structMap. It would be good to replace the CONTENTIDS with FILEID in the AIP METS file once processing is complete, unless the CONTENTIDS are in `<area>` or `<mptr>` elements, (if possible).

**Comment by evelynPM** [comment]
@ross-spencer just to clarify, do all custom structMaps now need to have the CONTENTIDS attribute in order to be parsed to the AIP METS file?

**Comment by evelynPM** [comment]
Ross and I discussed this and agreed that it's ok to require CONTENTIDS in custom structMaps, so long as the requirement is documented. Also, in the AIP METS file the CONTENTIDS don't appear in the Archivematica default structMap, only in the custom structMaps, so they can be left in.

**Comment by ross-spencer** [comment]
Thanks @evelynPM. I've moved this out of `In progress` to `Review` so we're a bit closer to `Verified` cc. @sallain or we can put to `Verified` but we'll need to update the docs with a little more information. The commit for the PID declaration method was added [here](https://github.com/artefactual/archivematica-docs/commit/8d25f6fa8982a5031a02aebf31ace7371885ac8d).

**Comment by sallain** [comment]
@ross-spencer Thanks! I think we've low-key decided that docs shouldn't block releases so I've added a docs tag to this - I've been doing this for all the 1.10 things that need docs updates. If you or @evelynPM want to add a separate docs issue, that might be good too. So this can be moved to Done if it is done!

**Comment by ross-spencer** [comment]
Hi @evelynPM just one more thing, you asked about this [here](https://github.com/archivematica/Issues/issues/283#issuecomment-511606025) and I proposed an addition to the script that I think would be of benefit [here](https://github.com/archivematica/Issues/issues/283#issuecomment-511762282). Should we still follow up on this while we have a chance with `1.10`? cc. @sallain as a reminder to follow-up on for today's AM weekly.

**Comment by sallain** [comment]
Needs to go in release notes!

**Comment by evelynPM** [comment]
@ross-spencer and I had a conversation about this & agreed that the SIP should fail if lack of CONTENTIDS in the custom structMap result in the structMap not being parsed to the AIP METS file.

**Comment by replaceafill** [comment]
@ross-spencer @evelynPM I'm code reviewing the PR connected to this issue. I've tried the `SingleObjectExample` and `MinimalStructMapExample` sample transfers but cannot see any failures in the process. I do see the `Custom structmap validated correctly` output in the `Verify mets_structmap.xml compliance` job.

Could we update the issue with a **Steps to reproduce** section? Thanks!

**Comment by ross-spencer** [comment]
Hi @replaceafill it's probably not that clear but the additional PR is to capture error conditions we were perhaps being a little too flexible with in the feature's first version which is only being introduced with `1.10`.

I've included four new structmaps into the fixtures to test against, so if you look at the xml filter [here](https://github.com/artefactual/archivematica/pull/1455/files?file-filters%5B%5D=.xml) you'll get an idea what you want to see Archivematica report, or fail on. 

In summary, if you want to take, say, `MinimalStructMapExample`, you need to modify it to four different states and transfer if four times:

1. No `CONTENTID` attributes (each of these will be found in `mets:fptr` elements or `mets:area` elements.
2. One missing `CONTENTID` attribute.
3. A add a path that does not exist to a `CONTENTID` attribute. 
4. `CONTENTID` attributes exist but are `""` empty. 

Crucially it means the feature will not fail silently, and its harder to put rogue data in, so we also want to make sure the sample data passes, I've tested  each of the four samples I created in the sample data repo and they all work as expected but you might want to verify.

**Comment by replaceafill** [comment]
Thank you @ross-spencer. I understand now what those sample transfers do.

I know the issue has back and forth discussion about the feature and provides samples and explanations, but I still think a **Steps to reproduce** section in each issue is worth for people less experienced in some of these features like me :blush:

**Comment by sallain** [comment]
@evelynPM I wonder if the docs need updating because of the late changes? It would be this section here: [Importing structural metadata with mets_structmap.xml](https://github.com/artefactual/archivematica-docs/blob/1.10/user-manual/transfer/import-metadata.rst#importing-structural-metadata-with-mets_structmapxml). The referenced scripts are [here](https://github.com/artefactual/archivematica-docs/tree/1.10/user-manual/transfer/scripts)

**Comment by ross-spencer** [comment]
Hi @sallain we didn't modify the mechanism with the last changes, so any changes would be focused on nuances about working with this feature, otherwise 🤞 we should be g2g?

**Comment by sallain** [comment]
@ross-spencer Just confirming, the structmap samples are still correct?

**Comment by ross-spencer** [comment]
@sallain yep, I would have submitted new ones with the changes otherwise. :+1:
